File: app/src/main/java/com/omar/retromp3recorder/app/shared/usecase/StopPlaybackAndRecordUC.java
package com.omar.retromp3recorder.app.shared.usecase;

import com.omar.retromp3recorder.app.playback.player.AudioPlayer;
import com.omar.retromp3recorder.app.recording.recorder.VoiceRecorder;

import javax.inject.Inject;

import io.reactivex.Completable;

public class StopPlaybackAndRecordUC implements NoParamsUseCase {

    private final AudioPlayer audioPlayer;
    private final VoiceRecorder voiceRecorder;

    //region constructor
    @Inject
    public StopPlaybackAndRecordUC(
            AudioPlayer audioPlayer,
            VoiceRecorder voiceRecorder
    ) {
        this.audioPlayer = audioPlayer;
        this.voiceRecorder = voiceRecorder;
    }
    //endregion

    public Completable execute() {
        return Completable
                .fromAction(() -> {
                    voiceRecorder.stopRecord();
                    audioPlayer.playerStop();
                });
    }
}

File: usecases/src/main/java/com/omar/retromp3recorder/bl/audio/StopPlaybackAndRecordUC.kt
package com.omar.retromp3recorder.bl.audio

import com.omar.retromp3recorder.audioplayer.AudioPlayer
import io.reactivex.rxjava3.core.Completable
import javax.inject.Inject

class StopPlaybackAndRecordUC @Inject constructor(
    private val audioPlayer: AudioPlayer,
    private val stopRecordUC: StopRecordUC,
    private val stateMapper: AudioStateMapper,
) {
    fun execute(): Completable = stateMapper
        .observe()
        .take(1)
        .flatMapCompletable { state: AudioState ->
            when (state) {
                AudioState.Playing,
                AudioState.Seek_Paused -> Completable.fromAction { audioPlayer.onInput(AudioPlayer.Input.Stop) }
                AudioState.Recording -> stopRecordUC.execute()
                is AudioState.Idle -> Completable.complete()
            }
        }
}
